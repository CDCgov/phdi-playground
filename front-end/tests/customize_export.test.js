import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import "@testing-library/jest-dom";
import CustomizeExportPage from '../app/customize_export/page'; // Adjust the import path as needed

// Mock the DataContext module
jest.mock('../utils/DataContext', () => ({
    useData: () => ({
        data: {
            "processed_values": {
                "message": "Parsing succeeded!",
                "parsed_values": {
                    "patient_id": "95a52bd2-2a17-4888-af52-be3472c94e0e",
                    "person_id": null,
                    "last_name": "BEAKER",
                    "first_name": "CATEST",
                    "birth_date": "2016-01-27",
                    "gender": "female"
                }
            }
        },
    }),
}));

global.URL.createObjectURL = jest.fn();
global.URL.revokeObjectURL = jest.fn();

describe("Customize Export Page", () => {
    it("Customize Export Page component should render page", () => {
        expect(render(<CustomizeExportPage />)).toMatchSnapshot();
    });
  
    it('should toggle checkbox when clicked', async () => {
      const user = userEvent.setup()
      render(<CustomizeExportPage />);
      
      // Find the "Deselect All" button and click it
      const deselectButton = screen.getByText('Deselect All');      
      await user.click(deselectButton);
  
      // Get all the checkboxes within the "checkbox-container" div
      const checkboxes = screen.getAllByTestId('checkbox');
  
      // Assert that all checkboxes are unchecked
      checkboxes.forEach((checkbox) => {
        expect(checkbox.querySelector('input')).not.toBeChecked();
      });
  
      // Click the "Select All" button to check boxes again
      const selectButton = screen.getByText('Select All');
      await user.click(selectButton);
  
      // Assert that all checkboxes are checked again
      checkboxes.forEach((checkbox) => {
        expect(checkbox.querySelector('input')).toBeChecked();
      });
    });
  });
  

test('Customize Export component should render and trigger download', async () => {
    const user = userEvent.setup()
    render(<CustomizeExportPage />);

    // Assert that the component renders
    // Mock the anchor click function to prevent actual navigation
    const anchorClickMock = jest.fn();
    const originalCreateElement = document.createElement;
    document.createElement = (tagName) => {
        if (tagName === 'a') {
            return {
                click: anchorClickMock,
                setAttribute: jest.fn(),
                removeAttribute: jest.fn(),
            };
        }
        return originalCreateElement(tagName);
    };

    // Find the "Export" button and click it
    const exportButton = screen.getByText('Download (.json)');
    await user.click(exportButton);

    // Assert that the anchor click function was called
    expect(anchorClickMock).toHaveBeenCalled();
});
