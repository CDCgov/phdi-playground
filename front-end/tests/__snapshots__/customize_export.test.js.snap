// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Customize Export Page Customize Export Page component should render page 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h1>
        Customize your bundle
      </h1>
      <h3>
        Select which fields you wish to 
        <i>
          <b>
            omit
          </b>
        </i>
         from the options below:
      </h3>
      <div
        class="section-container usa-table"
      >
        <div
          class="deselect-all-container"
        >
          <h3>
            Patient Information
          </h3>
           
          <button
            class="deselect-all-button usa-link"
          >
            Deselect All
          </button>
        </div>
        <div
          class="checkbox-container"
        >
          <div>
            <div
              class="usa-checkbox"
              data-testid="checkbox"
            >
              <input
                class="usa-checkbox__input"
                id="checkbox-patient_id"
                name="patient_id"
                type="checkbox"
              />
              <label
                class="usa-checkbox__label"
                for="checkbox-patient_id"
              >
                Patient Id
              </label>
            </div>
          </div>
          <div>
            <div
              class="usa-checkbox"
              data-testid="checkbox"
            >
              <input
                class="usa-checkbox__input"
                id="checkbox-person_id"
                name="person_id"
                type="checkbox"
              />
              <label
                class="usa-checkbox__label"
                for="checkbox-person_id"
              >
                Person Id
              </label>
            </div>
          </div>
          <div>
            <div
              class="usa-checkbox"
              data-testid="checkbox"
            >
              <input
                class="usa-checkbox__input"
                id="checkbox-last_name"
                name="last_name"
                type="checkbox"
              />
              <label
                class="usa-checkbox__label"
                for="checkbox-last_name"
              >
                Last Name
              </label>
            </div>
          </div>
          <div>
            <div
              class="usa-checkbox"
              data-testid="checkbox"
            >
              <input
                class="usa-checkbox__input"
                id="checkbox-first_name"
                name="first_name"
                type="checkbox"
              />
              <label
                class="usa-checkbox__label"
                for="checkbox-first_name"
              >
                First Name
              </label>
            </div>
          </div>
          <div>
            <div
              class="usa-checkbox"
              data-testid="checkbox"
            >
              <input
                class="usa-checkbox__input"
                id="checkbox-birth_date"
                name="birth_date"
                type="checkbox"
              />
              <label
                class="usa-checkbox__label"
                for="checkbox-birth_date"
              >
                Birth Date
              </label>
            </div>
          </div>
          <div>
            <div
              class="usa-checkbox"
              data-testid="checkbox"
            >
              <input
                class="usa-checkbox__input"
                id="checkbox-gender"
                name="gender"
                type="checkbox"
              />
              <label
                class="usa-checkbox__label"
                for="checkbox-gender"
              >
                Gender
              </label>
            </div>
          </div>
        </div>
        <br />
      </div>
      <button
        class="usa-button"
        data-testid="button"
        type="button"
      >
        Download (.json)
      </button>
      <button
        class="usa-button usa-button--outline usa-button"
        data-testid="button"
        type="button"
      >
        Cancel
      </button>
    </div>
  </body>,
  "container": <div>
    <h1>
      Customize your bundle
    </h1>
    <h3>
      Select which fields you wish to 
      <i>
        <b>
          omit
        </b>
      </i>
       from the options below:
    </h3>
    <div
      class="section-container usa-table"
    >
      <div
        class="deselect-all-container"
      >
        <h3>
          Patient Information
        </h3>
         
        <button
          class="deselect-all-button usa-link"
        >
          Deselect All
        </button>
      </div>
      <div
        class="checkbox-container"
      >
        <div>
          <div
            class="usa-checkbox"
            data-testid="checkbox"
          >
            <input
              class="usa-checkbox__input"
              id="checkbox-patient_id"
              name="patient_id"
              type="checkbox"
            />
            <label
              class="usa-checkbox__label"
              for="checkbox-patient_id"
            >
              Patient Id
            </label>
          </div>
        </div>
        <div>
          <div
            class="usa-checkbox"
            data-testid="checkbox"
          >
            <input
              class="usa-checkbox__input"
              id="checkbox-person_id"
              name="person_id"
              type="checkbox"
            />
            <label
              class="usa-checkbox__label"
              for="checkbox-person_id"
            >
              Person Id
            </label>
          </div>
        </div>
        <div>
          <div
            class="usa-checkbox"
            data-testid="checkbox"
          >
            <input
              class="usa-checkbox__input"
              id="checkbox-last_name"
              name="last_name"
              type="checkbox"
            />
            <label
              class="usa-checkbox__label"
              for="checkbox-last_name"
            >
              Last Name
            </label>
          </div>
        </div>
        <div>
          <div
            class="usa-checkbox"
            data-testid="checkbox"
          >
            <input
              class="usa-checkbox__input"
              id="checkbox-first_name"
              name="first_name"
              type="checkbox"
            />
            <label
              class="usa-checkbox__label"
              for="checkbox-first_name"
            >
              First Name
            </label>
          </div>
        </div>
        <div>
          <div
            class="usa-checkbox"
            data-testid="checkbox"
          >
            <input
              class="usa-checkbox__input"
              id="checkbox-birth_date"
              name="birth_date"
              type="checkbox"
            />
            <label
              class="usa-checkbox__label"
              for="checkbox-birth_date"
            >
              Birth Date
            </label>
          </div>
        </div>
        <div>
          <div
            class="usa-checkbox"
            data-testid="checkbox"
          >
            <input
              class="usa-checkbox__input"
              id="checkbox-gender"
              name="gender"
              type="checkbox"
            />
            <label
              class="usa-checkbox__label"
              for="checkbox-gender"
            >
              Gender
            </label>
          </div>
        </div>
      </div>
      <br />
    </div>
    <button
      class="usa-button"
      data-testid="button"
      type="button"
    >
      Download (.json)
    </button>
    <button
      class="usa-button usa-button--outline usa-button"
      data-testid="button"
      type="button"
    >
      Cancel
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
